kyo-prelude/shared/src/main/scala/kyo2/kernel/Trace.scala

```diff
@@ -0,0 +1,103 @@
+package kyo2.kernel
+
+import internal.*
+import java.util.Arrays
+import java.util.concurrent.atomic.AtomicInteger
+import scala.annotation.tailrec
+import scala.util.control.NoStackTrace
+
+class Trace(
+    private[kernel] var frames: Array[Frame],
+    private[kernel] var index: Int
+)
+
+object Trace:
+
+    private[kernel] def init: Trace = Trace(new Array[Frame](maxTraceFrames), 0)
+
+    abstract class Owner extends TracePool.Local:
```

Comment:
This is used as the super class of `Safepoint` to "flatten" the memory layout and avoid pointer chasing given how hot this code is.